<!-- end auto-generated rule header -->
<!-- Do not manually modify this header. Run: `pnpm eslint-docs` -->

## Rule details

(Detailed explanation of what the rule does, why it exists in the context of roblox-ts, and potential pitfalls it prevents.)

## Examples

Examples of **incorrect** code for this rule:

```js
// Code that violates the rule
const badExample = () => {
  // ...
};
```

Examples of **correct** code for this rule:

```js
// Code that adheres to the rule
const goodExample = () => {
  // ...
};
```

## Options

(If the rule has options, describe them here using a table or list. Otherwise,
remove this section.)

<!-- begin auto-generated rule options list -->
<!-- end auto-generated rule options list -->

Example configuration with options:

```json
{
	"roblox-ts/{{RULE_NAME}}": ["error", { /* optionName: optionValue */ }]
}
```

## Further Reading

(Optional: Links to relevant roblox-ts documentation, GitHub issues, or related concepts.)
